### ISSUE TITLE
S2187 FP: when Test methods are declared in nested types

### ISSUE THREAD

**Issue by costin-zaharia-sonarsource** [SonarSource/sonar-dotnet#5732]
### Description

Rule S2187, is raising a false positive when Test methods are declared in nested types

### Repro steps

```
    [TestFixture]
    public class A // Noncompliant, FP - In NUnit this scenario is valid.
    {
        public class Nested : A
        {
            [Test]
            public void Test()
            {
            }
        }
    }
```

### Expected behavior

The issue should not be raised.

### Actual behavior

S2187 is raised.

### Known workarounds

Declare the attribute on the nested type. This might be inconvenient when there are multiple nested types.

### Related information

* C#/VB.NET Plugins version: 8.40

**Comment by manfred-brands** [comment]
> [TestFixture]
    public class A // Noncompliant, FP - In NUnit this scenario is valid.

This is actually not a valid scenario.
If there are no tests NUnit will create an instance of `A` only to find out there are no tests to run.

There is a [ticket](https://github.com/nunit/nunit.analyzers/issues/840) in NUnit.Analyzers to warn about this use case.
The associated codefix will mark the class `A` as `abstract` because on its own it is not a `TestFixture`.

**Comment by martin-strecker-sonarsource** [comment]
Thank you @manfred-brands for bringing this up. We already have an exception in S2187:

> **Abstract classes** 
To facilitate the creation of common test cases, test logic, or test infrastructure, it is advisable to use a base class.
Additionally, in both NUnit and MSTest, abstract classes that are annotated with their respective attributes (TestFixture in NUnit and TestClass in MSTest) are automatically ignored.
Therefore, there is no need to raise an issue in this particular scenario.

Given that this exception exists in our rule and your recommendation also is to mark such base classes as `abstract` to avoid duplicate test executions, I will close this issue.

The reasoning is that you can change the declaration to `public abstract class A` and S2187 will no longer be raised. Also `abstract` prevents any potential test executions for the abstract base class. Such tests would be executed inside the derived test classes only, which can be useful for tests that are depended on overridden functionality.
